File: ReactNative/IMSDK/react-native-tim-js/android/src/main/java/com/reactnativetimjs/manager/ConversationManager.java
package com.reactnativetimjs.manager;

import com.reactnativetimjs.util.CommonUtils;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReadableMap;

import com.tencent.imsdk.v2.V2TIMCallback;
import com.tencent.imsdk.v2.V2TIMConversation;
import com.tencent.imsdk.v2.V2TIMConversationListener;
import com.tencent.imsdk.v2.V2TIMConversationResult;
import com.tencent.imsdk.v2.V2TIMManager;
import com.tencent.imsdk.v2.V2TIMValueCallback;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;

public class ConversationManager {
    private static V2TIMConversationListener conversationListener;

    public void removeConversationListener(Promise promise, ReadableMap arguments) {
        V2TIMManager.getConversationManager().removeConversationListener(conversationListener);
        CommonUtils.returnSuccess(promise, "removeConversationListener is done");
    }

    public void addConversationListener(Promise promise, ReadableMap arguments) {
        conversationListener = new V2TIMConversationListener() {
            @Override
            public void onSyncServerStart() {
                makeConversationListenerEventData("onSyncServerStart", null);
            }

            @Override
            public void onSyncServerFinish() {
                makeConversationListenerEventData("onSyncServerFinish", null);
            }

            @Override
            public void onSyncServerFailed() {
                makeConversationListenerEventData("onSyncServerFailed", null);
            }

            @Override
            public void onNewConversation(List<V2TIMConversation> conversationList) {
                LinkedList<HashMap<String, Object>> list = new LinkedList<HashMap<String, Object>>();
                for (int i = 0; i < conversationList.size(); i++) {
                    list.add(CommonUtils.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                makeConversationListenerEventData("onNewConversation", list);
            }

            @Override
            public void onConversationChanged(List<V2TIMConversation> conversationList) {
                LinkedList<HashMap<String, Object>> list = new LinkedList<HashMap<String, Object>>();
                for (int i = 0; i < conversationList.size(); i++) {
                    list.add(CommonUtils.convertV2TIMConversationToMap(conversationList.get(i)));
                }
                makeConversationListenerEventData("onConversationChanged", list);
            }

            @Override
            public void onTotalUnreadMessageCountChanged(long totalUnreadCount) {
                makeConversationListenerEventData("onTotalUnreadMessageCountChanged", totalUnreadCount);
            }
        };
        V2TIMManager.getConversationManager().addConversationListener(conversationListener);
        CommonUtils.returnSuccess(promise, "addConversationListener success");
    }

    public void getConversation(Promise promise, ReadableMap arguments) {
        String conversationID = arguments.getString("conversationID");
        V2TIMManager.getConversationManager().getConversation(conversationID,
                new V2TIMValueCallback<V2TIMConversation>() {
                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                    }

                    @Override
                    public void onSuccess(V2TIMConversation v2TIMConversation) {
                        CommonUtils.returnSuccess(promise,
                                CommonUtils.convertV2TIMConversationToMap(v2TIMConversation));
                    }
                });
    }

    private <T> void makeConversationListenerEventData(String eventType, T data) {
        CommonUtils.emmitEvent("conversationListener", eventType, data);
    }

    public void getConversationList(Promise promise, ReadableMap arguments) {
        String nextSeq = arguments.getString("nextSeq");
        int count = arguments.getInt("count");
        V2TIMManager.getConversationManager().getConversationList(Long.parseLong(nextSeq), count,
                new V2TIMValueCallback<V2TIMConversationResult>() {
                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                    }

                    @Override
                    public void onSuccess(V2TIMConversationResult v2TIMConversationResult) {
                        CommonUtils.returnSuccess(promise,
                                CommonUtils.convertV2TIMConversationResultToMap(v2TIMConversationResult));
                    }
                });
    }

    public void getConversationListByConversaionIds(Promise promise, ReadableMap arguments) {
        List<String> conversationIDList = CommonUtils
                .convertReadableArrayToListString(arguments.getArray("conversationIDList"));
        V2TIMManager.getConversationManager().getConversationList(conversationIDList,
                new V2TIMValueCallback<List<V2TIMConversation>>() {
                    @Override
                    public void onSuccess(List<V2TIMConversation> v2TIMConversations) {
                        LinkedList<HashMap<String, Object>> list = new LinkedList<HashMap<String, Object>>();
                        for (int i = 0; i < v2TIMConversations.size(); i++) {
                            list.add(CommonUtils.convertV2TIMConversationToMap(v2TIMConversations.get(i)));
                        }
                        CommonUtils.returnSuccess(promise, list);
                    }

                    @Override
                    public void onError(int i, String s) {
                        CommonUtils.returnError(promise, i, s);
                    }
                });
    }

    public void pinConversation(Promise promise, ReadableMap arguments) {
        String conversationID = arguments.getString("conversationID");
        boolean isPinned = arguments.getBoolean("isPinned");
        V2TIMManager.getConversationManager().pinConversation(conversationID, isPinned, new V2TIMCallback() {
            @Override
            public void onSuccess() {
                CommonUtils.returnSuccess(promise, null);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }
        });
    }

    public void getTotalUnreadMessageCount(Promise promise, ReadableMap arguments) {
        V2TIMManager.getConversationManager().getTotalUnreadMessageCount(new V2TIMValueCallback<Long>() {
            @Override
            public void onSuccess(Long aLong) {
                CommonUtils.returnSuccess(promise, aLong);
            }

            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }
        });
    }

    public void deleteConversation(Promise promise, ReadableMap arguments) {
        // session id
        String conversationID = arguments.getString("conversationID");
        V2TIMManager.getConversationManager().deleteConversation(conversationID, new V2TIMCallback() {
            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }

            @Override
            public void onSuccess() {
                CommonUtils.returnSuccess(promise, null);
            }
        });

    }

    public void setConversationDraft(Promise promise, ReadableMap arguments) {
        // session id
        String conversationID = arguments.getString("conversationID");
        String draftText = arguments.getString("draftText");
        if (draftText == "") {
            draftText = null;
        }
        V2TIMManager.getConversationManager().setConversationDraft(conversationID, draftText, new V2TIMCallback() {
            @Override
            public void onError(int i, String s) {
                CommonUtils.returnError(promise, i, s);
            }

            @Override
            public void onSuccess() {
                CommonUtils.returnSuccess(promise, null);
            }
        });

    }
}

File: ReactNative/IMSDK/react-native-tim-js/ios/manager/conversationManager.swift
import Foundation
import ImSDK_Plus

class ConversationManager {
	private var conversationListener: ConversationListener = ConversationListener();

	public func getConversationList(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		if let nextSeq = param["nextSeq"] as? String,
		   let count = param["count"] as? Int32 {
            V2TIMManager.sharedInstance().getConversationList(UInt64(nextSeq) ?? 0 , count: count, succ: {
				conversations, nextSeq, finished in
				
                let dict = V2ConversationResultEntity.init(conversations: conversations ?? [], nextSeq: String(nextSeq), finished: finished).getDict();
				CommonUtils.resultSuccess(method: "getConversationList", resolve: resolve, data: dict);
			}, fail: TencentImUtils.returnErrorClosures(method: "getConversationList", resolve: resolve))
		}
	}
	
	public func getConversationListByConversaionIds(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		if let conversationIDList =  param["conversationIDList"] as? [String] {
		   V2TIMManager.sharedInstance().getConversationList(conversationIDList, succ: {
               if($0 != nil){
                   let dict = $0!.map { V2ConversationEntity.getDict(info: $0) };
                   CommonUtils.resultSuccess(method: "getConversationListByConversaionIds", resolve: resolve, data: dict);
               }
				
			}, fail: TencentImUtils.returnErrorClosures(method: "getConversationListByConversaionIds", resolve: resolve))
		}
	}
	
	public func getConversation(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		if let conversationID = param["conversationID"] as? String {
			V2TIMManager.sharedInstance().getConversation(conversationID, succ: {
				conversation in
				
				CommonUtils.resultSuccess(method: "getConversation", resolve: resolve, data: V2ConversationEntity.getDict(info: conversation!));
			}, fail: TencentImUtils.returnErrorClosures(method: "getConversation", resolve: resolve))
		}
	}
	
	public func deleteConversation(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		let conversationID = param["conversationID"] as? String;
		
		V2TIMManager.sharedInstance()?.deleteConversation(conversationID, succ: {
			() -> Void in
			
			CommonUtils.resultSuccess(method: "deleteConversation", resolve: resolve)
		}, fail: TencentImUtils.returnErrorClosures(method: "deleteConversation", resolve: resolve));
	}
	
	public func setConversationDraft(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		let conversationID = param["conversationID"] as? String;
		let draftText = param["draftText"] as? String;
		
		V2TIMManager.sharedInstance()?.setConversationDraft(conversationID, draftText: draftText, succ: {
			() -> Void in
			
			CommonUtils.resultSuccess(method: "setConversationDraft", resolve: resolve)
		}, fail: TencentImUtils.returnErrorClosures(method: "setConversationDraft", resolve: resolve))
	}
	
	public func pinConversation(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		let conversationID = param["conversationID"] as? String;
		let isPinned = param["isPinned"] as? Bool;
		
		V2TIMManager.sharedInstance()?.pinConversation(conversationID, isPinned: isPinned ?? false, succ: {
			() -> Void in
			CommonUtils.resultSuccess(method: "pinConversation", resolve: resolve, data: "ok")
		}, fail: TencentImUtils.returnErrorClosures(method: "pinConversation", resolve: resolve))
	}
	
	public func getTotalUnreadMessageCount(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		V2TIMManager.sharedInstance()?.getTotalUnreadMessageCount({
			totalCount -> Void in
			CommonUtils.resultSuccess(method: "getTotalUnreadMessageCount", resolve: resolve, data: totalCount)
		}, fail: TencentImUtils.returnErrorClosures(method: "getTotalUnreadMessageCount", resolve: resolve))
	}

	public func addConversationListener(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
        V2TIMManager.sharedInstance().addConversationListener(listener: conversationListener);
		CommonUtils.resultSuccess(method: "addConversationListener", resolve: resolve, data: "addConversationListener is done");
	}

	public func removeConversationListener(param: [String: Any], resolve:@escaping RCTPromiseResolveBlock) {
		V2TIMManager.sharedInstance().removeConversationListener(listener: conversationListener);
		CommonUtils.resultSuccess(method: "removeConversationListener", resolve: resolve, data: "removeConversationListener is done");
	}
}
